src/workerd/api/web-socket.c++

```diff
@@ -310,8 +340,16 @@ kj::Promise<DeferredProxy<void>> WebSocket::couple(kj::Own<kj::WebSocket> other)
 
   auto& context = IoContext::current();
 
-  auto upstream = other->pumpTo(*self);
-  auto downstream = self->pumpTo(*other);
+  auto upstream = other->pumpTo(*self).catch_([](kj::Exception&& e) {
+    KJ_UNWRAP_OR(WebSocketProtocolError::fromException(e), {
```

Comment:
I'm actually not sure if it matters all that since this is pretty much passing through the error, but you might also want to use `kj::throwFatalException(e)` instead. I believe that it basically extends our async stack a bit?

Reply:
btw, can this be written to use coroutine syntax rather than using `catch_`?

And yes, this should use `kj::throwFatalException(e)` instead.

Reply:
Ah, that's how you say "this Maybe is empty". I'd found `KJ_IF_MAYBE(_, m){} else { ... }`, but that's ugly.

(In my ideal world, `kj::Maybe` would provide `operator bool()` for "has stuff in it", or maybe a `.empty()` method, but this is not that world.)

I fixed everything but the coroutine syntax. I can convert each individual `pumpTo` call, but currently the two promises get joined down below and I'm not sure what to do with that. Suggestions welcome.

Reply:
@jasnell I've coroutine-ified things as much as I can. There's still a call to `kj::joinPromises` that I don't see how to get rid of, but there's no more `catch_`. Does this look good to you?